echopype/calibrate/ek80_complex.py

```diff
@@ -273,47 +285,53 @@ def compress_pulse(backscatter: xr.DataArray, chirp: Dict) -> xr.DataArray:
     xr.DataArray
         A data array containing pulse compression output.
     """
-    pc_all = []
-
-    for chan in backscatter["channel"]:
-        # Select channel `chan` and drop the specific beam dimension if all of the values are nan.
-        backscatter_chan = backscatter.sel(channel=chan).dropna(dim="beam", how="all")
-
-        # Create NaN mask
-        # If `backscatter_chan` is lazy loaded, then `nan_mask` too will be lazy loaded.
-        nan_mask = np.isnan(backscatter_chan)
-
-        # Zero out backscatter NaN values
-        # If `nan_mask` is lazy loaded, then resulting `backscatter_chan` will be lazy loaded.
-        backscatter_chan = xr.where(nan_mask, 0.0 + 0j, backscatter_chan)
-
-        # Extract transmit values
-        tx = chirp[str(chan.values)]
-
+    # Select channel `chan` and drop the specific beam dimension if all of the values are nan.
+    # Additionally, in the same for loop, compute the replica dictionary values from the chirp.
+    backscatter_NaN_beam_drop_all = []
+    replica_dict = {}
+    for channel in backscatter["channel"]:
+        # TODO: Once `dropna` allows for dropping along multiple dimensions, put this outside of the
+        # loop and remove the concatenate.
+        backscatter_NaN_beam_drop = backscatter.sel(channel=channel).dropna(dim="beam", how="all")
+        backscatter_NaN_beam_drop_all.append(backscatter_NaN_beam_drop)
+
+        # Extract tx
+        tx = chirp[str(channel.values)]
         # Compute complex conjugate of transmit values and reverse order of elements
-        replica = np.flipud(np.conj(tx))
+        replica_dict[str(channel.values)] = np.flipud(np.conj(tx))
+    # Concatenate backscatter channels with dropped NaN beam dimensions.
+    backscatter_NaN_beam_drop_all = xr.concat(backscatter_NaN_beam_drop_all, dim="channel")
```

Comment:
Question: I think the original intention to dropna with `dim="beam"` is to remove the padded sections along the `range_sample` dimension, since some channels contain data that are shorter. So, if you dropna for a particular channel, but other channels are longer, wouldn't this `xr.concat` recreate the dropped NaN values?

Reply:
Oh yeah, I think you're right. Let me think and tinker with this.

Reply:
CODE:

```
data = np.array([
    [np.nan, np.nan], [4, 5], [np.nan, 6], [8, 9]
])
da = xr.DataArray(
    data,
    dims=["channel", "beam"],
    coords={
        "channel": np.arange(4),
        "beam": np.arange(2)
    })
dropped_nan_beam_da_list = []
for channel_index in range(len(da["channel"])):
    dropped_da = da.isel(channel=channel_index).dropna("beam", how="all")
    print(f"`dropped_da` shape (chan idx {channel_index}): {dropped_da.shape[0]}")
    dropped_nan_beam_da_list.append(dropped_da)
print("concatenated da:", xr.concat(dropped_nan_beam_da_list, dim="channel"))
```

OUTPUT:

```
`dropped_da` shape (chan idx 0): 0
`dropped_da` shape (chan idx 1): 2
`dropped_da` shape (chan idx 2): 1
`dropped_da` shape (chan idx 3): 2
concatenated da: <xarray.DataArray (channel: 4, beam: 2)> Size: 64B
array([[nan, nan],
       [ 4.,  5.],
       [nan,  6.],
       [ 8.,  9.]])
Coordinates:
  * beam     (beam) int64 16B 0 1
  * channel  (channel) int64 32B 0 1 2 3
```

Reply:
So yeah you're right, this padding does happen. Let me tinker with the implementation that's currently in this PR. There's probably a clever way skip to skip this concat.

Reply:
Oh, I think your code above is missing the `range_sample` dimension, which is where that padding happens.

But yeah, I think the easiest is to just not dropna here. We have 2 options:
- try drop the NaNs in the `_convolve_per_channel` function, since there you do have a loop that is channel by channel. The only thing is that you have to drop the NaN and add them back to ensure the output length is the same
- just pay the computational price

Maybe you could try seeing how the speed differ for these 2 approaches? It is possible that with all the optimization people have put into convolution, the speed difference is small (and hopefully the memory consumption is reasonable too).

Reply:
That sounds good, I'll try it out. I have a suspicion that the 2nd option of just not dropping NaNs might be faster, since the time needed to convolve on all NaNs may be faster than the time needed to remove and add back NaNs.

Reply:
Yeah, I agree, it very possibly would be that way!